apps/ex_wire/lib/ex_wire/struct/warp_queue.ex

```diff
@@ -0,0 +1,275 @@
+          block_data,
+          trie,
+          warp_queue.current_block_number,
+          warp_queue.current_block_hash
+        )
+
+      next_block_tree = Blocktree.update_best_block(block_tree, block)
+      next_warp_queue = %{warp_queue | unprocessed_block_data: updated_unprocessed_block_data}
+
+      process_completed_blocks(
+        next_warp_queue,
+        next_block_tree,
+        next_trie
+      )
+    else
+      {warp_queue, block_tree, trie}
+    end
+  end
+
+  # Note: we haven't connected state hashes, so we're not there yet
+  @spec process_block(BlockData.t(), Trie.t(), integer(), EVM.hash()) :: {Block.t(), Trie.t()}
+  defp process_block(block_data, trie, block_number, parent_hash) do
+    # First, we need to validate some aspect of the block, that's currently
+    # ommited.
+
+    transactions_hash = get_transactions_or_receipts_root(block_data.header.transactions_rlp)
+    ommers_hash = get_ommers_hash(block_data.header.ommers_rlp)
+    receipts_root = get_transactions_or_receipts_root(block_data.receipts_rlp)
+
+    header = %Header{
+      parent_hash: parent_hash,
+      ommers_hash: ommers_hash,
+      beneficiary: block_data.header.author,
+      state_root: block_data.header.state_root,
+      transactions_root: transactions_hash,
+      receipts_root: receipts_root,
+      logs_bloom: block_data.header.logs_bloom,
+      difficulty: block_data.header.difficulty,
+      number: block_number,
+      gas_limit: block_data.header.gas_limit,
+      gas_used: block_data.header.gas_used,
+      timestamp: block_data.header.timestamp,
+      extra_data: block_data.header.extra_data,
+      mix_hash: block_data.header.mix_hash,
+      nonce: block_data.header.nonce
+    }
+
+    block = %Block{
+      block_hash: Header.hash(header),
+      header: header,
+      transactions: block_data.header.transactions,
+      receipts: block_data.receipts,
+      ommers: block_data.header.ommers
+    }
+
+    Exth.inspect(block, "Processing Block!")
+
+    # Store the block to our db
+    {:ok, {_, next_trie}} = Block.put_block(block, trie)
+
+    # TODO: Store each transaction to our db
+    # {subtrie, updated_trie} =
+    #   TrieStorage.update_subtrie_key(
+    #     trie,
+    #     block.header.transactions_root,
+    #     ExRLP.encode(i),
+    #     encoded_transaction
+    #   )
+
+    # TODO: Store each receipt
+    # {subtrie, updated_trie} =
+    #   TrieStorage.update_subtrie_key(
+    #     trie,
+    #     block.header.receipts_root,
+    #     ExRLP.encode(i),
+    #     encoded_receipt
+    #   )
+
+    # Mark our new block head
+    {
+      block,
+      next_trie
+    }
+  end
+
+  # Tries to get the transaction or receipts root by encoding the transaction trie
+  @spec get_transactions_or_receipts_root([ExRLP.t()]) :: MerklePatriciaTree.Trie.root_hash()
+  def get_transactions_or_receipts_root(transactions_or_receipts_rlp) do
+    # this is a throw-away
+    db = MerklePatriciaTree.Test.random_ets_db()
```

Comment:
why test in-memory database is used? Also, I think we should use `TrieStorage` behaviour to working with database

Reply:
If we can sure, the idea is simply turning a list of transactions into the state root of a trie with all of the transactions included. We **never** store this value to disk (hence why we use ets and this discard it).